File: week3/binary-search.java
class Solution {
    public int search(int[] nums, int target) {
        return search(nums, 0, nums.length - 1, target);
    }
    
    public int search(int[] nums, int lo, int hi, int target) {
        if (lo > hi) {
            return -1;
        }
        int mid = lo + (hi - lo) / 2;
        if (nums[mid] == target) {
            return mid;
        } else if (nums[mid] > target) {
            return search(nums, lo, mid - 1, target);
        } else {
            return search(nums, mid + 1, hi, target);
        }
    }
}

File: summer2021/unit3/binary-search.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        return self.binary_search(target, 0, len(nums) - 1, nums)
    
    def binary_search(self, target, lo, hi, nums):
        mid = lo + int((hi - lo) / 2)
        if lo > hi:
            return -1
        elif nums[mid] == target:
            return mid
        elif target < nums[mid]:
            return self.binary_search(target, 0, mid - 1, nums)
        else:
            return self.binary_search(target, mid + 1, hi, nums)